editor/rendering/tiles.ts

```diff
@@ -0,0 +1,517 @@
+            Layer.createFromRawData(this.tile_size, [])
+        );
+    }
+
+
+    /// Draw operation only used by a rendering context.
+    draw(gl: WebGLRenderingContext, program: Program, camera: Camera) {
+
+        // Init dynamic_layers if needed
+        this.initDynamicLayers(gl);
+
+        // Update content of buffers if needed
+        this.updateBuffers(camera);
+
+        if (!this.vertex_linked) {
+            this.vertex_linked = new BufferLinkedToProgram(
+                program,
+                this.vertex_buffer,
+                'tile_pos'
+            );
+        }
+
+        for (let layer of this.dynamic_layers) {
+            layer.draw(gl, program, this.vertex_linked, this.index_buffer);
+        }
+    }
+
+    private initDynamicLayers(gl: WebGLRenderingContext) {
+        if (this.dynamic_layers.length === 0) {
+            for (let sl of this.static_layers) {
+                this.dynamic_layers.push(
+                    Layer.createForRendering(gl, sl)
+                );
+            }
+        }
+    }
+
+    private updateBuffers(camera: Camera) {
+
+        // Check if the update is needed or not.
+        let new_ij: [number, number] = [0, 0];
+        let new_hw: [number, number] = [0, 0];
+        let f = CFP * this.tile_size;
+
+        let diff: [number, number] = [
+            camera.pos[0] - this.pos[0],
+            camera.pos[1] - this.pos[1]
+        ];
+
+        // Compute the new position of the camera in tile space
+        new_ij[0] = Math.min(
+            Math.max(Math.ceil(diff[1] / f), 0) * CFP,
+            this.height
+        );
+        new_ij[1] = Math.min(
+            Math.max(Math.ceil(diff[0] / f), 0) * CFP,
+            this.width
+        );
+        new_hw[0] = Math.max(
+            Math.min(Math.floor((diff[1] + camera.hos) / f) * CFP, this.height),
+            0
+        );
+        new_hw[1] = Math.max(
+            Math.min(Math.floor((diff[0] + camera.wos) / f) * CFP, this.width),
+            0
+        );
+
+
+        if (new_ij[0] !== this.old_cam_ij[0] || new_ij[1] !== this.old_cam_ij[1]
+         || new_hw[0] !== this.old_cam_hw[0] || new_hw[1] !== this.old_cam_hw[1]
+         || this.is_dirty)
+        {
+            this.old_cam_ij = new_ij;
+            this.old_cam_hw = new_hw;
+            this.is_dirty = false;
+
+            let tiles_w = new_hw[0] - new_ij[0];
+            let tiles_h = new_hw[1] - new_ij[1];
+            let nb_tiles = tiles_w * tiles_h;
+            let indices = new Uint16Array(nb_tiles * 6);
+            let vertices = new Float32Array(nb_tiles * 8);
+
+            // Vertex buffer
+            let index = 0;
+            let indices_index = 0;
+            let eps = 0.1;
+            for (let i = new_ij[0]; i < new_hw[0]; ++i) {
+                for (let j = new_ij[1]; j < new_hw[1]; ++j) {
+                    let first_index = index / 2;
+
+                    vertices[index++] = this.pos[0] + 2 * this.tile_size * j - eps;
```

Comment:
This `2` here is a hack. The rendered size is incorrect without it. I haven't find a good explanation for it. It just doesn't make sense.